staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1338,3 +1477,110 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
+	for _, v := range crd.Spec.Versions {
+		decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+	}
+
+	appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+		ref := metav1.OwnerReference{
+			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+			Name:       crd.Name,
+			UID:        crd.UID,
+		}
+		for _, r := range sv.OwnerReferences {
+			if r == ref {
+				return nil
+			}
+		}
+		sv.OwnerReferences = append(sv.OwnerReferences, ref)
+		return nil
+	}
+	return storageversion.UpdateStorageVersionFor(
+		ctx,
+		r.storageVersionClient,
+		r.apiserverID,
+		gr,
+		encodingVersion,
+		decodableVersions,
+		appendOwnerRefFunc)
+}
+
+// getOrCreateStorageVersionUpdateQueue_locked returns the channel for the given UID,
+// or create a new one and a new goroutine if necessary. The goroutine keeps processing
+// requests until the channel is closed.
+// NOTE: caller should hold the customStorageLock until the write to the returned channel is finished,
+// to make sure the storage version update is in sync with the underlying storage change.
+func (r *crdHandler) getOrCreateStorageVersionUpdateQueue_locked(uid types.UID) chan *storageVersionUpdateRequest {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
+		return queue.q
+	}
+
+	queue := make(chan *storageVersionUpdateRequest, maxStorageVersionUpdateRequests)
+	ctx, cancel := context.WithCancel(context.TODO())
+	r.storageVersionUpdateQueues[uid] = &storageVersionUpdateQueue{
+		q:      queue,
+		cancel: cancel,
+	}
+	go func() {
+		for req := range queue {
+			if req.wait != nil {
+				<-req.wait
+			}
+			if err := r.updateStorageVersionFor(ctx, req.crd); err != nil {
+				utilruntime.HandleError(err)
+			}
+			if req.processed != nil {
+				select {
+				case <-ctx.Done():
+					// The queue was cancelled. Potentially we didn't finish this
+					// storage version update request.
+					close(req.abandoned)
+				default:
+					close(req.processed)
+				}
+			}
+		}
+	}()
+	return queue
+}
+
+// stopStorageVersionUpdateQueue_locked close the channel for the given UID.
+// Closing the channel ensures that we don't leak goroutine for the given
+// UID.
+func (r *crdHandler) stopStorageVersionUpdateQueue_locked(uid types.UID) {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
+		// Drain the channel before closing it, so that if the CRD is re-created,
+		// the old requests won't race with the new requests. We can safely discard
+		// the old storage version update requests because:
+		//   1. if the GroupResource is deleted forever, all CRs will be
+		//      GC'ed and the storage version doesn't matter.
+		//   2. if the CRD gets deleted and re-created in a short time and the old
+		//      CRs remain, the new CRD will update new storage version
+		klog.V(4).Infof("Cancelling the storage version update queue for CRD with UID %v.", uid)
+		queue.cancel()
+		for len(queue.q) > 0 {
```

Comment:
If you take my previous comment, "Why not checking the ctx.Done() at the beginning?", then you don't need to drain the queue here.

Reply:
See https://github.com/kubernetes/kubernetes/pull/96403#discussion_r553047824. IMO, even if we check the ctx.Done() at the beginning, draining the queue actively still has its value compared to relying on `ctx.Done()` checks passively.

Reply:
Can you elaborate on the value of doing the draining here?
Also, since there are two consumers of the queue, do you need a lock before checking `len(queue.q)`?

Reply:
Relying on `ctx.Done()` checks means the queued requests will be processed and failed after the currently being-processed request hit one of the `ctx.Done()` checks. On the other hand, draining the queue allows us to fail the queued requests and release the resources immediately.

> there are two consumers of the queue
Nice catch! I added an unblocking select to handle the race.

Reply:
> Relying on ctx.Done() checks means the queued requests will be processed and failed after the currently being-processed request hit one of the ctx.Done() checks.

That's right, but I don't think it's much difference in practice, especially if you add a check of `ctx.Done()` at the beginning of the other consumer code. I would sacrifice a little performance for easier concurrency model.